operatorcourier/flatten.py

```diff
@@ -0,0 +1,131 @@
+import logging
+import os
+import re
+from yaml import safe_load
+import operatorcourier.identify as identify
+
+logger = logging.getLogger(__name__)
+
+
+def get_flattened_files(source_dir: str) -> list:
+    """
+    :param source_dir: the directory containing different versions
+    of operator bundles (CRD, CSV, package) in separate version directories
+    :return: a list of files needed to be copied to the flattened directory
+    """
+
+    # extract package file and version folders from source_dir
+    file_names, folder_names = [], []
+    for item in os.listdir(source_dir):
+        item_path = f'{source_dir}/{item}'
+        if os.path.isfile(item_path):
+            file_names.append(item)
+        elif os.path.isdir(item_path):
+            folder_names.append(item)
+        else:
+            logger.warning("Ignoring %s as it is neither a folder "
+                           "nor a regular file.", item)
+
+    file_paths_to_copy = []
+    crd_dict = {}  # { CRD_NAME => (VERSION, CRD_PATH) }
+
+    # parse each version folder and parse CRD, CSV files
+    for version_folder_name in folder_names:
+        if not semantic_version_string(version_folder_name):
+            logger.warning('Ignoring %s as the folder name '
+                           'is not a valid semantic version.', version_folder_name)
+            continue
+
+        logger.info('Parsing folder: %s...', version_folder_name)
+
+        contains_csv = False
+        for item in os.listdir(f'{source_dir}/{version_folder_name}'):
+            version_folder_path = f'{source_dir}/{version_folder_name}'
+            item_path = f'{version_folder_path}/{item}'
+
+            if not os.path.isfile(item_path):
+                logger.warning('Ignoring %s as it is not a regular file.', item)
+
+            with open(item_path, 'r') as f:
+                file_content = f.read()
+
+            yaml_type = identify.get_operator_artifact_type(file_content)
+
+            if yaml_type == 'ClusterServiceVersion':
+                contains_csv = True
+                file_paths_to_copy.append(item_path)
+            elif yaml_type == 'CustomResourceDefinition':
+                crd_name = safe_load(file_content)['metadata']['name']
```

Comment:
Don't you want to raise here `OpCourier*` errors instead `ParseError/KeyError`

Reply:
+1